File: src/core/cpp/entropy.cpp
#include "entropy.h"

#include <algorithm>
#include <cmath>
#include <stdexcept>

namespace sentinel {

double EntropyCalculator::calculate(const uint8_t* data, size_t length) {
    if (data == nullptr && length > 0) {
        throw std::invalid_argument("data pointer cannot be null when length > 0");
    }
    if (length == 0) {
        return 0.0;
    }

    unsigned int frequencies[256] = {0};
    for (size_t index = 0; index < length; ++index) {
        frequencies[data[index]]++;
    }

    double entropy = 0.0;
    for (unsigned int frequency : frequencies) {
        if (frequency == 0) {
            continue;
        }
        const double probability = static_cast<double>(frequency) / static_cast<double>(length);
        entropy -= probability * std::log2(probability);
    }

    return entropy;
}

std::vector<double> EntropyCalculator::calculate_windowed(
    const uint8_t* data,
    size_t length,
    size_t window_size
) {
    if (window_size == 0) {
        throw std::invalid_argument("window_size must be greater than zero");
    }

    std::vector<double> scores;
    if (length == 0) {
        scores.push_back(0.0);
        return scores;
    }
    if (length < window_size) {
        scores.push_back(calculate(data, length));
        return scores;
    }

    for (size_t offset = 0; offset < length; offset += window_size) {
        const size_t chunk_size = std::min(window_size, length - offset);
        scores.push_back(calculate(data + offset, chunk_size));
    }

    return scores;
}

std::vector<uint8_t> EntropyCalculator::entropy_heatmap(
    const uint8_t* data,
    size_t length,
    size_t window_size
) {
    if (window_size == 0) {
        throw std::invalid_argument("window_size must be greater than zero");
    }

    std::vector<uint8_t> heatmap;
    if (length == 0) {
        heatmap.push_back(0);
        return heatmap;
    }

    for (size_t offset = 0; offset < length; offset += window_size) {
        const size_t chunk_size = std::min(window_size, length - offset);
        const double entropy = calculate(data + offset, chunk_size);
        const auto intensity = static_cast<uint8_t>(std::min(255.0, (entropy / 8.0) * 255.0));
        heatmap.push_back(intensity);
    }

    return heatmap;
}

}  // namespace sentinel

File: src/modules/entropy.py
"""Shannon entropy analysis helpers."""

from __future__ import annotations

import math
from collections import Counter
from typing import Any

from src.core.python.exceptions import EntropyCalculationError


class ShannonEntropyCalculator:
    """
    Calculate Shannon entropy for binary data.

    H(X) = -Σ P(x) * log2(P(x))
    """

    @staticmethod
    def calculate(data: bytes) -> float:
        """Calculate Shannon entropy of byte sequence."""
        if not data:
            return 0.0

        byte_counts = Counter(data)
        total_bytes = len(data)
        entropy = 0.0

        for count in byte_counts.values():
            probability = count / total_bytes
            entropy -= probability * math.log2(probability)

        return entropy

    @staticmethod
    def calculate_windowed(data: bytes, window_size: int = 1024) -> tuple[float, list[float]]:
        """Calculate average entropy and per-window entropy scores."""
        if window_size <= 0:
            raise EntropyCalculationError("window_size must be greater than zero")

        if len(data) < window_size:
            avg = ShannonEntropyCalculator.calculate(data)
            return avg, [avg]

        scores: list[float] = []
        for index in range(0, len(data), window_size):
            window = data[index : index + window_size]
            scores.append(ShannonEntropyCalculator.calculate(window))

        return sum(scores) / len(scores), scores

    @staticmethod
    def entropy_heatmap(data: bytes, window_size: int = 256) -> list[dict[str, Any]]:
        """Generate per-window entropy metadata for visualization."""
        if window_size <= 0:
            raise EntropyCalculationError("window_size must be greater than zero")

        heatmap: list[dict[str, Any]] = []
        for index in range(0, len(data), window_size):
            window = data[index : index + window_size]
            entropy = ShannonEntropyCalculator.calculate(window)
            if entropy >= 7.0:
                color = "red"
            elif entropy >= 5.0:
                color = "yellow"
            else:
                color = "green"

            heatmap.append({"position": index, "entropy": entropy, "color": color})
        return heatmap
